src/validation.rs

```diff
@@ -0,0 +1,147 @@
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use crate::{Commitment, Error, Hash, KeyImage, KeyManager, PublicKey, Result, SpentProof};
+use blst_ringct::ringct::RingCtTransaction;
+use blsttc::Signature;
+use std::collections::{BTreeMap, BTreeSet};
+
+// Here we are putting transaction validation logic that is common to both
+// MintNode::reissue() and Dbc::confirm_valid().
+//
+// It is best to have the validation logic in one place only!
+//
+// Note also that MintNode is server-side (mint) and Dbc is client-side (wallet).
+// In a future refactor, we intend to break the code into 3 modules:
+//   server, client, and common.  (or maybe mint, wallet, and common)
+// So TransactionValidator would go into common.
+//
+// Another way to do this would be to create a NewType wrapper for RingCtTransaction.
+// We can discuss if that is better or not.
+
+pub struct TransactionValidator {}
+
+impl TransactionValidator {
+    // note: for spent_proofs to validate, the mint_verifier must have/know the spentbook section's public key.
+    pub fn validate<K: KeyManager>(
+        mint_verifier: &K,
+        transaction: &RingCtTransaction,
+        transaction_sigs: &BTreeMap<KeyImage, (PublicKey, Signature)>,
+        spent_proofs: &BTreeSet<SpentProof>,
+    ) -> Result<(), Error> {
+        // Do quick checks first to reduce potential DOS vectors.
+
+        if transaction_sigs.len() < transaction.mlsags.len() {
+            return Err(Error::MissingSignatureForInput);
+        }
+
+        let tx_hash = Hash::from(transaction.hash());
+
+        // Verify that each input has a corresponding valid mint signature.
+        for (key_image, (mint_key, mint_sig)) in transaction_sigs.iter() {
+            if !transaction
+                .mlsags
+                .iter()
+                .any(|m| m.key_image == *key_image.as_ref())
+            {
+                return Err(Error::UnknownInput);
+            }
+
+            mint_verifier
+                .verify(&tx_hash, mint_key, mint_sig)
+                .map_err(|e| Error::Signing(e.to_string()))?;
+        }
+
+        Self::validate_without_sigs_internal(mint_verifier, transaction, tx_hash, spent_proofs)
+    }
+
+    pub fn validate_without_sigs<K: KeyManager>(
+        mint_verifier: &K,
+        transaction: &RingCtTransaction,
+        spent_proofs: &BTreeSet<SpentProof>,
+    ) -> Result<(), Error> {
+        let tx_hash = Hash::from(transaction.hash());
+        Self::validate_without_sigs_internal(mint_verifier, transaction, tx_hash, spent_proofs)
+    }
+
+    fn validate_without_sigs_internal<K: KeyManager>(
+        mint_verifier: &K,
+        transaction: &RingCtTransaction,
+        transaction_hash: Hash,
+        spent_proofs: &BTreeSet<SpentProof>,
+    ) -> Result<(), Error> {
+        if transaction.mlsags.is_empty() {
+            return Err(Error::TransactionMustHaveAnInput);
```

Comment:
This check should be moved to `RingCtTransaction::verify()` if it's not already there.

Reply:
created #128